**Bryan Quah, Matthew Xu** <br />
CSE 415 Winter 2018 Project <br />
*Notebook to investigate the data we are dealing with and to determine how to approach it.*

In [103]:
import numpy as np
def aggregate_cifar():
    ''' 
    Aggregates the CIFAR-10 data and returns a single array consisting of 50000 arrays
    with RGB values for each image.
    '''
    full_batch = []
    full_labels = []
    FILE_NAME = "cifar-10-batches-py/data_batch_"
    FILE_NUM = 5 
    IMAGES_PER_BATCH = 10000
    for i in range(FILE_NUM):
        batch_data, labels_data = unpickle(FILE_NAME + str(i + 1)) 
        for j in range(IMAGES_PER_BATCH):
            full_batch.append(np.append(batch_data[j], labels_data[j]))
    return full_batch

def unpickle(file):
    ''' 
    Unpacks the data files for CIFAR-10
    '''
    # TODO:
    # return the list of labels as well
    import cPickle
    with open(file, 'rb') as fo: 
        dict = cPickle.load(fo)
    return dict['data'], dict['labels']


In [106]:
len(aggregate_cifar()[0])

3073

In [84]:
unpickle("cifar-10-batches-py/data_batch_1")

(array([[ 59,  43,  50, ..., 140,  84,  72],
        [154, 126, 105, ..., 139, 142, 144],
        [255, 253, 253, ...,  83,  83,  84],
        ..., 
        [ 71,  60,  74, ...,  68,  69,  68],
        [250, 254, 211, ..., 215, 255, 254],
        [ 62,  61,  60, ..., 130, 130, 131]], dtype=uint8),
 [6,
  9,
  9,
  4,
  1,
  1,
  2,
  7,
  8,
  3,
  4,
  7,
  7,
  2,
  9,
  9,
  9,
  3,
  2,
  6,
  4,
  3,
  6,
  6,
  2,
  6,
  3,
  5,
  4,
  0,
  0,
  9,
  1,
  3,
  4,
  0,
  3,
  7,
  3,
  3,
  5,
  2,
  2,
  7,
  1,
  1,
  1,
  2,
  2,
  0,
  9,
  5,
  7,
  9,
  2,
  2,
  5,
  2,
  4,
  3,
  1,
  1,
  8,
  2,
  1,
  1,
  4,
  9,
  7,
  8,
  5,
  9,
  6,
  7,
  3,
  1,
  9,
  0,
  3,
  1,
  3,
  5,
  4,
  5,
  7,
  7,
  4,
  7,
  9,
  4,
  2,
  3,
  8,
  0,
  1,
  6,
  1,
  1,
  4,
  1,
  8,
  3,
  9,
  6,
  6,
  1,
  8,
  5,
  2,
  9,
  9,
  8,
  1,
  7,
  7,
  0,
  0,
  6,
  9,
  1,
  2,
  2,
  9,
  2,
  6,
  6,
  1,
  9,
  5,
  0,
  4,
  7,
  6,
  7,
  1,
  8,
  1,
  1,
  2,
  8,
 

In [79]:
len(unpickle("cifar-10-batches-py/data_batch_5"))

10000

In [82]:
len(aggregate_cifar())

50000

In [13]:
print(data['data'][9999]) # dict with keys: batch_label, data, filesnames, labels
print(len(data['data']))
print("------")
print(data['data'])
data['data']

[ 62  61  60 ..., 130 130 131]
10000
------
[[ 59  43  50 ..., 140  84  72]
 [154 126 105 ..., 139 142 144]
 [255 253 253 ...,  83  83  84]
 ..., 
 [ 71  60  74 ...,  68  69  68]
 [250 254 211 ..., 215 255 254]
 [ 62  61  60 ..., 130 130 131]]


array([[ 59,  43,  50, ..., 140,  84,  72],
       [154, 126, 105, ..., 139, 142, 144],
       [255, 253, 253, ...,  83,  83,  84],
       ..., 
       [ 71,  60,  74, ...,  68,  69,  68],
       [250, 254, 211, ..., 215, 255, 254],
       [ 62,  61,  60, ..., 130, 130, 131]], dtype=uint8)

In [7]:
len(data['data'][0]) # first 1024 entries contain red vals, then green, then blue

3072

In [10]:
len(data['data']) # total number of images

10000

# Notes

#### CIFAR-10 Notes:
- There are 5 data_batch files for training classifiers
- Use test_batch file to test training of the classifier and report it

#### Random Forest/Decision Tree Notes:
- Question nodes should split images according to rgb value of each pixel, i.e. possible 256^3072 questions
- There should be 10 leaf nodes for each label/class - airplane, automobile, bird, etc.

# Progress

#### (a) What are the main techniques you are featuring in your project?
For building a decision tree in the random forest, we will construct a tree consisting of question nodes and terminal nodes. The question nodes will be used to determine where each row of the data should be directed - where each row either passes or fails a question node, creating at most two children nodes. As such a decision tree will also be a binary tree. From there we will create multiple decision trees to create a forest of which each tree will "vote" for what label should be assigned to the object/data.

#### (b) How far along is your implementation (what works, what does not yet work?)
Most of the time has been spent studying how to correctly implement our classifiers - **Random Forests and Neutal Networks**. We currently have boilerplate code to implement a Decision Tree which will later be used for the Random Forest classifier. We are looking into Random Forests to understand how each decision tree should be different from one another.